src/factor_graph.h

```diff
@@ -175,8 +173,17 @@ class CompactFactorGraph {
 };
 
 inline double CompactFactorGraph::potential(
-    const CompactFactor& factor, const variable_value_t assignments[]) {
-  return factor.potential(vifs.get(), assignments, -1, -1);
+    const CompactFactor& factor, const variable_value_t assignments[],
+    factor_value_t factor_value) {
+  // For boolean, this is stored in CompactFactor.value;
+  //     caller shouldn't set the factor_value arg.
+  // For categorical, this is stored in Factor.factor_params.value.
+  //     caller should pass in via the factor_value arg.
+  // TODO: better data structures than the factor_params duct tape...
+  if (factor_value == Factor::DEFAULT_VALUE) {
+    factor_value = factor.value;
+  }
```

Comment:
switch-casing on `factor.func_id` might be clearer? and perhaps the multiplication can be folded into `CompactFactor::potential(_,_,_,_)`?